tree/tree.go

```diff
@@ -0,0 +1,723 @@
+type KeyMap struct {
+	Down   key.Binding
+	Up     key.Binding
+	Toggle key.Binding
+
+	// Help toggle keybindings.
+	ShowFullHelp  key.Binding
+	CloseFullHelp key.Binding
+
+	Quit key.Binding
+}
+
+// DefaultKeyMap is the default set of key bindings for navigating and acting
+// upon the tree.
+var DefaultKeyMap = KeyMap{
+	Down: key.NewBinding(
+		key.WithKeys("down", "j", "ctrl+n"),
+		key.WithHelp("↓/j", "down"),
+	),
+	Up: key.NewBinding(
+		key.WithKeys("up", "k", "ctrl+p"),
+		key.WithHelp("↑/k", "up"),
+	),
+	Toggle: key.NewBinding(
+		key.WithKeys("enter"),
+		key.WithHelp("⏎", "toggle"),
+	),
+
+	// Toggle help.
+	ShowFullHelp: key.NewBinding(
+		key.WithKeys("?"),
+		key.WithHelp("?", "more"),
+	),
+	CloseFullHelp: key.NewBinding(
+		key.WithKeys("?"),
+		key.WithHelp("?", "close help"),
+	),
+
+	Quit: key.NewBinding(
+		key.WithKeys("q", "ctrl+c"),
+		key.WithHelp("q", "quit"),
+	),
+}
+
+// Model is the Bubble Tea model for this tree element.
+type Model struct {
+	showHelp bool
+	// OpenCharacter is the character used to represent an open node.
+	OpenCharacter string
+	// ClosedCharacter is the character used to represent a closed node.
+	ClosedCharacter string
+	// KeyMap encodes the keybindings recognized by the widget.
+	KeyMap KeyMap
+	// styles sets the styling for the tree
+	styles Styles
+	Help   help.Model
+
+	// Additional key mappings for the short and full help views. This allows
+	// you to add additional key mappings to the help menu without
+	// re-implementing the help component. Of course, you can also disable the
+	// list's help component and implement a new one if you need more
+	// flexibility.
+	AdditionalShortHelpKeys func() []key.Binding
+	AdditionalFullHelpKeys  func() []key.Binding
+
+	root   *Node
+	width  int
+	height int
+	// yOffset is the vertical offset of the selected node.
+	yOffset int
+}
+
+// Node is a a node in the tree
+// Node implements lipgloss's tree.Node
+type Node struct {
+	// tree is used as the renderer layer
+	tree *ltree.Tree
+
+	// yOffset is the vertical offset of the selected node.
+	yOffset int
+
+	// depth is the depth of the node in the tree
+	depth int
+
+	// isRoot is true for every Node which was added with tree.Root
+	isRoot bool
+	open   bool
+
+	// value is the root value of the node
+	value any
+
+	// TODO: expose a getter for this in lipgloss?
+	rootStyle lipgloss.Style
+
+	opts *itemOptions
+
+	// TODO: move to lipgloss.Tree?
+	size int
+}
+
+// IsSelected returns whether this item is selected.
+func (t *Node) IsSelected() bool {
+	return t.yOffset == t.opts.treeYOffset
+}
+
+// Depth returns the depth of the node in the tree.
+func (t *Node) Depth() int {
+	return t.depth
+}
+
+// Size returns the number of nodes in the tree.
+// Note that if a child isn't open, its size is 1
+func (t *Node) Size() int {
+	return t.size
+}
+
+// YOffset returns the vertical offset of the Node
+func (t *Node) YOffset() int {
+	return t.yOffset
+}
```

Comment:
These can be used to allow for more flexible styling.
I found the `StyleFunc` of `lipgloss.Tree` doesn't support some stuff I want, like appending a symbol to the end of each item at the same indentation to the right of it. Maybe there are more use cases.

Reply:
Have you seen the [toggle example](https://github.com/charmbracelet/lipgloss/blob/master/examples/tree/toggle/main.go#L40-L53)? I actually put that one together in order to make sure `tree` can work interactively (though I of course may have missed some things).